protobuf-format.md

```diff
@@ -0,0 +1,181 @@
+# Protobuf Event Format for CloudEvents - Version 1.0
+
+## Abstract
+
+[Protocol Buffers][proto-home] is a mechanism for marshalling structured data.
+This document defines how CloudEvents are represented using [version 3][proto-3]
+of that specification.
+
+In this document the terms *Protocol Buffers*, *protobuf*, and *proto* are used
+interchangeably.
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Data](#3-data)
+4. [Transport](#4-transport)
+5. [Examples](#5-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce] is a standardized and protocol-agnostic definition of the
+structure and metadata description of events. This specification defines how the
+elements defined in the CloudEvents specification are are represented using
+a protobuf schema. The  [schema][proto-schema] defines an envolope/body
+construct similar to that prescriped in the [JSON format][json-format].
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvents attributes for use as protobuf message
+properties.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for protobuf, as such this
+specification uses 'application/x-protobuf' to identify such content.
+
+## 2. Type System
+
+The CloudEvents type system is mapped to protobuf as follows :
+
+| CloudEvents   | protobuf |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][proto-scalars] |
+| Integer       | [sint32][proto-scalars] |
+| String        | [string][proto-scalars] |
+| Binary        | [bytes][proto-scalars] |
+| URI           | [string][proto-scalars] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][proto-scalars] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][proto-timestamp]  |
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the protobuf
+[schema][proto-schema].
+
+Required and Optional attributes are represented explicitly as protobuf fields.
+
+CloudEvent extension attributes are represented using a map construct enabling
+direct support of the CloudEvent [type system][ce-types].
+
+```proto
+map<string, CloudEventAttribute> extensions = 1;
```

Comment:
This map is labeled `extensions`. Are the core attributes handled the same way, or by unique tags?

The example at the end suggests that known attributes are assigned unique tags.

Reply:
originally I had all the defined attributes in a _map_ structure, I flipped to a model when the defined attributes are explicit (as in the JSON schema).

So in that model only extension attributes need to be carried in the _map_.

Reply:
My concern is about what happens when core attributes are added, e.g. in a 1.1 spec.

Clients which were built before the 1.1 spec will map these attributes into the `extensions` map, while post-1.1 clients will map them into a unique tag-assigned field. Recipients of messages will then need to write complicated logic to map pre-1.1 messages to post-1.1 messages, losing most of the benefit of code auto-generation for protocol buffers.

Yes, this is a "perfect is the enemy of the good" case, but my experience has been that not planning for expansion tends to lead to "everything was good until we tried to touch something, and now it's all terrible."

Another acceptable scenario is to accept that we can't add further attributes to the 1.0 proto message definition, and all additional attributes (whether currently documented extensions or not-yet-defined extensions) must be stored in `extensions`.

The last option is to keep a global tag number registry mapping extension names to tags, but that only helps humans and doesn't really help conversion gateways, which would need to be able to load the list dynamically when transforming a message between proto and other encodings.

Reply:
Maybe there's a half-way house solution where all the 'required' attributes are explicitly define but the optional ones aren't, on the basis that we can't add any required attributes without changing the major version..